tcljava/doc/TclException.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: TclException.3,v 1.1 1998/10/14 21:09:08 cvsadmin Exp $
'\" 
.so man.macros
.TH TclException 3 1.0 Tcl "Jacl Library Procedures"
.BS
.SH EXCEPTION
tcl.lang.TclException -- This class is used to signal exception
conditions in the execution of Tcl scripts.

.SH CONSTRUCTORS
.sp
\fBTclException\fR(int \fIccode\fR)
.sp
\fBTclException\fR(Interp \fIinterp\fR, String \fImsg\fR, int \fIccode\fR)
.sp
\fBTclException\fR(Interp \fIinterp\fR, String \fImsg\fR)

.SH METHODS
.sp
int \fBgetCompletionCode\fR()

.SH ARGUMENTS
.AS Interp interp .
.AP Interp interp
Internal representation to be stored in newly created TclObject.
.AP String msg
Error message to store in the \fIinterp\fR's result object.
.AP int ccode
Completion code to store inside the TclException.
.BE

.SH DESCRIPTION
.TP
\fBgetCompletionCode\fR
The \fBgetCompletionCode\fR method is used to retrieve the completion
code of the command that threw the exception.  The following
are the four possible integer values for the completion code:
.RS
.TP
TCL.ERROR
The command couldn't be completed successfully;
the interpreter's result describes what went wrong.
.TP
TCL.RETURN
The command requests that the current procedure
return; the interpreter's result contains the
procedure's return value.
.TP
TCL.BREAK
The command requests that the innermost loop
be exited; the interpreter's result is meaningless.
.TP
TCL.CONTINUE
Go on to the next iteration of the current loop;
the interpreter's result is meaningless.
.RE
.TP
\fBTclException\fR(int \fIccode\fR)
This constructor creates a TclException with the completion code \fIccode\fR.
.TP
\fBTclException\fR(Interp \fIinterp\fR, String \fImsg\fR, int \fIccode\fR)
This constructor extends the above constructor.  If \fIinterp\fR is
non-null, its result object will be set to contain \fImsg\fR.
.TP
\fBTclException\fR(Interp \fIinterp\fR, String \fImsg\fR)
This constructor is the same as the above constructor, except that the
completion code is set to TCL.ERROR.

.SH "EQUIVALENT C FUNCTIONALITY"
The Exception handling syntax in the C API can be
represented by the following canonical C code:
.CS
	int foo(Tcl_Interp interp, ...)
	{
	    if (ok) {
		Tcl_SetResult(interp, result);
		return TCL_OK;
	    } else {
		Tcl_SetResult(interp, "err message ...");
		return TCL_ERROR;
	    }
	}
.CE
The return value of the foo() procedure is the completion code. TCL_OK
indicates a normal completion status, and all other values indicate an
exception. The above code can be written in Java as:
.CS
	void foo(Interp interp, ...)
	{
	    if (ok) {
		interp.setResult(result);
	    } else {
		interp.setResult("err message ...");
		throw new TclException(TCL.ERROR);
	    }
	}
.CE
In the Java API, foo()'s return type is void.  Rather than returning
the completion code, the completion code is either stored in a
TclException or implied to be OK.  If the foo() method completes
normally, no TclException is thrown, and the completion code is
implied to be OK.  Otherwise, an exception is thrown with the desired
completion code (TCL.ERROR in the example above).

.SH "SEE ALSO"
TclNumArgsException, TclRuntimeError, Interp

.SH KEYWORDS
error, exception, return, result


tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tcljava/tcl/lang/TclException.java
/*
 * TclException.java --
 *
 *	This file defines the TclException class used by Tcl to report
 *	generic script-level errors and exceptions.
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclException.java,v 1.2 2000/04/03 14:09:11 mo Exp $
 *
 */

package tcl.lang;

/*
 * TclException is used to interrupt the Tcl script currently being
 * interpreted by the Tcl Interpreter. Usually, a TclException is thrown
 * to indicate a script level error, e.g.:
 *
 *	- A syntax error occurred in a script.
 *	- A unknown variable is referenced.
 *	- A unknown command is executed.
 *	- A command is passed incorrected.
 *
 * A TclException can also be thrown by Tcl control structure commands such
 * as "return" and "continue" to change the flow of control in
 * a Tcl script.
 *
 * A TclException is accompanied by two pieces of information: the error
 * message and the completion code. The error message is a string stored in
 * the interpreter result. After a TclException is thrown and caught, the
 * error message can be queried by Interp.getResult().
 *
 * The completion code indicates why the TclException is generated. It is
 * stored in the compCode field of this class.
 */

public class TclException extends Exception {

/*
 * Stores the completion code of a TclException.
 */

private int compCode;

/*
 * An index that indicates where an error occurs inside a Tcl
 * string. This is used to add the offending command into the stack
 * trace.
 *
 * A negative value means the location of the index is unknown.
 *
 * Currently this field is used only by the Jacl interpreter.
 */

protected int errIndex;


/*
 *----------------------------------------------------------------------
 *
 * TclException --
 *
 *	Create an TclException with the given error message and
 *	completion code and indicate the location of the error in a
 *	script.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The instance fields are initialized; the message is assigned to
 *	the interpreter's result if interp is non-null.
 *
 *----------------------------------------------------------------------
 */

protected
TclException(
    Interp interp,		// Current interpreter. May be null if unknown.
    String msg,			// Error message.
    int ccode,			// Completion code.
    int idx) 			// Error index.
{
    super(msg);
    if (ccode == TCL.OK) {
	throw new TclRuntimeError(
		"The reserved completion code TCL.OK (0) cannot be used " +
		"in TclException");
    }
    compCode = ccode;
    errIndex = idx;

    if (interp != null && msg != null) {
	interp.setResult(msg);
    }
}

/*
 *----------------------------------------------------------------------
 *
 * TclException --
 *
 *	Create a TclException with the given completion code.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The instance fields are initialized.
 *
 *----------------------------------------------------------------------
 */

public
TclException(
    int ccode)			// Completion code.
{
    super();
    if (ccode == TCL.OK) {
	throw new TclRuntimeError(
		"The reserved completion code TCL.OK (0) cannot be used");
    }
    compCode = ccode;
    errIndex = -1;
}

/*
 *----------------------------------------------------------------------
 *
 * TclException --
 *
 *	Create an TclException with the given error message. The
 *	completion code is set to ERROR by default.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The instance fields are initialized; the message is assigned to
 *	the interpreter's result if interp is non-null.
 *
 *----------------------------------------------------------------------
 */

public
TclException(
    Interp interp,		// Current interpreter. May be null if unknown.
    String msg)			// Error message.
{
    this(interp, msg, TCL.ERROR, -1);
}

/*
 *----------------------------------------------------------------------
 *
 * TclException --
 *
 *	Create an TclException with the given error message and
 *	completion code.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The instance fields are initialized; the message is assigned to
 *	the interpreter's result if interp is non-null.
 *
 *----------------------------------------------------------------------
 */

public
TclException(
    Interp interp,		// Current interpreter. May be null if unknown.
    String msg,			// Error message.
    int ccode)			// Completion code.
{
    this(interp, msg, ccode, -1);
}

/*
 *----------------------------------------------------------------------
 *
 * getCompletionCode --
 *
 *	Returns the current completion code.
 *
 * Results:
 *	The int value of compCode.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

final public int 
getCompletionCode() 
{
    return compCode;
}

/*
 *----------------------------------------------------------------------
 *
 * setCompletionCode --
 *
 *	Sets the current completion code.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The completion code is changed.
 *
 *----------------------------------------------------------------------
 */

final void
setCompletionCode(
    int ccode)			// New completion code. 
{
    if (ccode == TCL.OK) {
	throw new TclRuntimeError(
		"The reserved completion code TCL.OK (0) cannot be used");
    }
    compCode = ccode;
}

} // end TclException

